src/render/gif.rs

```diff
@@ -0,0 +1,277 @@
+                let pixel_b = pixels.get((start + 2) as usize).unwrap_or(&0);
+                let pixel_a = pixels.get((start + 3) as usize).unwrap_or(&0);
+
+                //the value is multiplied by the alpha of said pixel
+                //the more the pixel is transparent, the less the pixel has an importance
+                colors[(pixel_r / 3 + pixel_g / 3 + pixel_b / 3) as usize] +=
+                    u32::from(pixel_a / 255);
+            }
+        }
+
+        let mut sum = 0;
+        for (color_value, count) in colors.iter().enumerate() {
+            sum += *count;
+
+            if u32::from(sum) >= num_pixels / 2 {
+                if color_value == 0 {
+                    return 1;
+                }
+                return color_value as u8;
+            }
+        }
+
+        1
+    }
+
+    pub fn read_frame(frame: &Frame, gif_height: i32, gif_width: i32) -> Vec<u8> {
+        let median_color = Self::calculate_median_color_value(frame, gif_height, gif_width);
+
+        let mut image = Vec::new();
+        let mut buf: u8 = 0;
+
+        //the u64 is just in case someone put a gif that's huge (in terms of
+        // resolution), it shouldn't break
+        let width = u64::from(frame.width);
+        let height = u64::from(frame.height);
+
+        let pixels = &frame.buffer;
+
+        for y in 0..gif_height {
+            //if y is outside of the gif width
+            if y as u64 >= height {
+                continue;
+            }
+
+            //if y is outside of the screen
+            if y >= DISPLAY_HEIGHT {
+                continue;
+            }
+            for x in 0..gif_width {
+                //since we're using an array of u8, every 8 bit we need to start with a new int
+                if x % 8 == 0 && x != 0 {
+                    image.push(buf);
+                    buf = 0;
+                }
+                //if x is outside of the gif width
+                if x as u64 >= width {
+                    continue;
+                }
+
+                //if x is outside of the screen
+                if x >= DISPLAY_WIDTH {
+                    continue;
+                }
+
+                //calculating the index
+                let start: u64 = ((y as u64) * width + (x as u64)) * 4;
+
+                //getting the value of the pixels
+                let pixel_r = pixels.get(start as usize).unwrap_or(&0);
+                let pixel_g = pixels.get((start + 1) as usize).unwrap_or(&0);
+                let pixel_b = pixels.get((start + 2) as usize).unwrap_or(&0);
+
+                let mean = pixel_r / 3 + pixel_g / 3 + pixel_b / 3;
+                //I'm not sure if we should do something with the alpha channel of the gif
+                //I decided not to, but maybe we should
+
+                if mean >= median_color {
+                    //which bit to turn on
+                    let shift = x % 8;
+                    buf = buf + (128 >> shift);
```

Comment:
`buf += (128 >> shift)`

Reply:
I thought I tried it and it didn't work, but it works now, maybe I did something wrong, will fix